### PR TITLE
Autocomplete search for Android Demo App

### PR DESCRIPTION
**Pull request by ianthetechie**
This makes the Android demo app into something more like a "normal" navigation experience with a single map view for everything and an autocomplete overlay. This involved... a few things ;)

* Making the map view usable without an active navigation session.
* Using the grid views for completely custom user overlays.
* Tip-toeing around a lot of very bad :tm: things in MapLibre Android wrt camera and user location tracking on init.
* Making the demo scene work with either simulated or live location providers

# Usage as a non-nav map

In the process, I had to address most of the same issues on Android that @hactar ran into with using this as a non-navigation map in #275. I tried a few ideas and ended up with a fairly similar approach (after accounting for Kotlin + Jetpack idioms) as that PR. Since I wasn't intending to do that by any means, I'm now pretty sure that the iOS PR is also going the right direction, so I'll figure out what else we need to do to get that over the line.

NOTE: I have _not_ yet deeply thought through the Jetpack approach to super deep customization like we do with chainable modifiers in SwiftUI, but I think this PR gets Android most of the way there, and it started from a significantly worse state than iOS.

# Outstanding tasks

- [x] Thoroughly test the "new install" flow (with permissions prompts)
- [ ] Decide on how to simulate navigation (button in the UI?), have the demo scene switch to the simulated provider during the simulated trip, and then switch back as soon as it's done.
- [x] The camera should switch to the navigating camera automatically when starting a trip; it's supposed to but I think I broke something

# Misc

While doing this, I realized that we can probably move the grid stuff to the compose playground (same thing for SwiftUI). This seems to be generally useful and I think we can move this outside Ferrostar and build on top of it with our "defaults" since it's all composable :)

### CODE REVIEW

**Review by Archdoog — COMMENTED** [review]
Looking pretty sweet 👍. A handful of deeper comments we probably need to discuss and get sorted pretty soon. Though we can merge this before we solve them, as this PR definitely improves things from main.

File: android/composeui/src/main/res/drawable/direction_u_turn.xml
[the remark this answers is not in the document]
Comment (ianthetechie) [reply android/composeui/src/main/res/drawable/direction_u_turn.xml:1]: Ahh yeah good call.

File: android/demo-app/src/main/java/com/stadiamaps/ferrostar/AppModule.kt [hunk 1]
```diff
@@ -65,11 +69,12 @@ object AppModule {
     FerrostarForegroundServiceManager(appContext, DefaultForegroundNotificationBuilder(appContext))
   }
 
+  // TODO: This is hard-coded for golf cart routing; change to something else before merging
```
[the remark this answers is not in the document]
Comment (ianthetechie) [reply android/demo-app/src/main/java/com/stadiamaps/ferrostar/AppModule.kt:72]: ha! Yeah, I'm open to that if we can do it without too much work. There's a fine line between a demo app and recreating Google Maps ;)

File: android/demo-app/src/main/java/com/stadiamaps/ferrostar/DemoNavigationScene.kt [hunk 2]
```diff
@@ -60,18 +73,40 @@ fun DemoNavigationScene(
...
+              // FIXME
+              //              locationProvider.addListener(locationUpdateListener, executor)
+            }
           }
           permissions.getOrDefault(Manifest.permission.ACCESS_COARSE_LOCATION, false) -> {
-            // TODO
+            // TODO: Probably alert the user that this is unusable for navigation
             // onAccess()
           }
+          // TODO: Foreground service permissions; we should block access until approved on API 34+
```
Comment (ianthetechie) [inline android/demo-app/src/main/java/com/stadiamaps/ferrostar/DemoNavigationScene.kt:109]: Self-marking for a re-read; this needs a bit of cleanup.

File: android/demo-app/src/main/java/com/stadiamaps/ferrostar/DemoNavigationScene.kt [hunk 3]
```diff
@@ -37,13 +46,17 @@ import uniffi.ferrostar.WaypointKind
...
 ) {
+  val executor = remember { Executors.newSingleThreadScheduledExecutor() }
+
   // Keeps the screen on at consistent brightness while this Composable is in the view hierarchy.
   KeepScreenOnDisposableEffect()
 
-  var viewModel by remember { mutableStateOf<NavigationViewModel?>(null) }
+  var viewModel by remember {
+    mutableStateOf<NavigationViewModel>(IdleNavigationViewModel(locationProvider))
+  }
```
[the remark this answers is not in the document]
Comment (ianthetechie) [reply android/demo-app/src/main/java/com/stadiamaps/ferrostar/DemoNavigationScene.kt:58]: Gotcha. Yeah, let's discuss tomorrow what the best way to do this is. I stall have a bit of a hard time thinking in the right terms on Android.

File: android/demo-app/src/main/java/com/stadiamaps/ferrostar/MainActivity.kt [hunk 4]
```diff
@@ -44,8 +43,8 @@ class MainActivity : ComponentActivity(), AndroidTtsStatusListener {
     AppModule.ferrostarCore.spokenInstructionObserver = AppModule.ttsObserver
 
     // Set up the location provider
-    AppModule.locationProvider.lastLocation = initialSimulatedLocation
-    AppModule.locationProvider.warpFactor = 2u
+    //    AppModule.locationProvider.lastLocation = initialSimulatedLocation
+    //    AppModule.locationProvider.warpFactor = 2u
```
Comment (ianthetechie) [inline android/demo-app/src/main/java/com/stadiamaps/ferrostar/MainActivity.kt:47]: The TODO here is to enable switching location providers.

File: android/maplibreui/src/main/java/com/stadiamaps/ferrostar/maplibreui/views/DynamicallyOrientingNavigationView.kt [hunk 5]
```diff
@@ -46,7 +47,8 @@ import com.stadiamaps.ferrostar.maplibreui.views.overlays.PortraitNavigationOver
  *   route line.
  * @param config The configuration for the navigation view.
  * @param onTapExit The callback to invoke when the exit button is tapped.
- * @param content Any additional composable map symbol content to render.
+ * @param userContent TODO docs
```
Comment (ianthetechie) [inline android/maplibreui/src/main/java/com/stadiamaps/ferrostar/maplibreui/views/DynamicallyOrientingNavigationView.kt:50]: TODO :P

File: android/core/src/main/java/com/stadiamaps/ferrostar/core/Location.kt [hunk 6]
```diff
@@ -78,13 +78,21 @@ class AndroidSystemLocationProvider(context: Context) : LocationProvider {
...
     listeners[listener] = androidListener
 
     val handler = Handler(Looper.getMainLooper())
 
     executor.execute {
       handler.post {
+        val last = locationManager.getLastKnownLocation(getBestProvider())
+        if (last != null) {
+          androidListener.onLocationChanged(last)
+        }
```
[the remark this answers is not in the document]
Comment (ianthetechie) [reply android/core/src/main/java/com/stadiamaps/ferrostar/core/Location.kt:95]: Yeah, it needs a `?`, but good call :)

File: android/demo-app/src/main/java/com/stadiamaps/ferrostar/AppModule.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (ianthetechie) [reply android/demo-app/src/main/java/com/stadiamaps/ferrostar/AppModule.kt:72]: We can throw in a segmented button here with walking, cycling, and driving and call it a day: https://developer.android.com/reference/kotlin/androidx/compose/material3/package-summary#SingleChoiceSegmentedButtonRow(androidx.compose.ui.Modifier,androidx.compose.ui.unit.Dp,kotlin.Function1)

File: android/maplibreui/src/main/java/com/stadiamaps/ferrostar/maplibreui/NavigationMapView.kt [hunk 7]
```diff
@@ -45,17 +48,21 @@ fun NavigationMapView(
     locationRequestProperties: LocationRequestProperties =
         LocationRequestProperties.NavigationDefault(),
     snapUserLocationToRoute: Boolean = true,
-    onMapReadyCallback: (Style) -> Unit = { camera.value = navigationCamera },
-    content: @Composable @MapLibreComposable() ((State<NavigationUiState>) -> Unit)? = null
+    onMapReadyCallback: (Style) -> Unit = {
+      if (viewModel.isNavigating()) camera.value = navigationCamera
```
[the remark this answers is not in the document]
Comment (ianthetechie) [reply android/maplibreui/src/main/java/com/stadiamaps/ferrostar/maplibreui/NavigationMapView.kt:52]: TL;DR, `onMapReady` is a useful hook, but we are abusing it :P This needs to trigger on ANY nav start, so we need a new hook for that (ex: `onNavigationStartCallback`) for major state changes.... the hooks probably belong at the view model level (ex: a method or interface which you implement that we hook up for you).

Other thought: maybe the same pattern works on iOS with the delegate.

File: android/demo-app/src/main/java/com/stadiamaps/ferrostar/MainActivity.kt (hunk 4 above)
Comment (ianthetechie) [reply android/demo-app/src/main/java/com/stadiamaps/ferrostar/MainActivity.kt:47]: Let's do another segmented control ;)

File: android/composeui/src/main/res/drawable/direction_u_turn.xml
[the remark this answers is not in the document]
Comment (ianthetechie) [reply android/composeui/src/main/res/drawable/direction_u_turn.xml:1]: This was broken, in a slightly different way ;) There is probably additional breakage, but we can test that later. Tracking in #302

File: android/maplibreui/src/main/java/com/stadiamaps/ferrostar/maplibreui/NavigationMapView.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (ianthetechie) [reply android/maplibreui/src/main/java/com/stadiamaps/ferrostar/maplibreui/NavigationMapView.kt:52]: I'm still struggling to resolve the best way to go about this sort of thing 😅 It seems like it's fairly elegant and not difficult to implement this locally wherever you need it. The complicating factor here is that our camera is currently _bidirectional_. In the specific case here, I think we would need the view model to actually "own" everything about the camera too (and provide methods for updating its state). I'm going to leave some comments to that effect and implement the navigation camera changes locally for the moment.

[the remark this answers is not in the document]
Comment (ianthetechie) [reply android/maplibreui/src/main/java/com/stadiamaps/ferrostar/maplibreui/NavigationMapView.kt:53]: Addressed in the last push.

**Comment by ianthetechie** [comment]
Discussed on today's call. We will merge with some minor cleanup and changing the default profile to something more common than golf carts for now and keep improving things. Stuff we still want to fix in the short term:

* The demo app should have its own view model eventually
* Switching location providers (to allow simulated routes without code changes)

We agreed it's best to merge early though since this definitely improves things a lot + lets devs test :)

**Pull request merged**